src/dag/put.js

```diff
@@ -0,0 +1,69 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const dagCBOR = require('ipld-dag-cbor')
+const promisify = require('promisify-es6')
+const CID = require('cids')
+const multihash = require('multihashes')
+const SendOneFile = require('../utils/send-one-file')
+
+function noop () {}
+
+module.exports = (send) => {
+  const sendOneFile = SendOneFile(send, 'dag/put')
+
+  return promisify((dagNode, options, callback) => {
+    if (typeof options === 'function') {
+      return setImmediate(() => callback(new Error('no options were passed')))
+    }
+
+    callback = callback || noop
+
+    let hashAlg = options.hash || 'sha2-256'
+    let format
+    let inputEnc
+
+    if (options.cid && CID.isCID(options.cid)) {
+      format = options.cid.codec
+      hashAlg = multihash.decode(options.cid.multihash).name
+      prepare()
+    } else if (options.format) {
+      format = options.format
+      prepare()
+    } else {
+      callback(new Error('Invalid arguments'))
+    }
+
+    function prepare () {
+      inputEnc = 'raw'
+
+      if (format === 'dag-cbor') {
+        dagCBOR.util.serialize(dagNode, finalize)
+      }
+      if (format === 'dag-pb') {
+        dagPB.util.serialize(dagNode, finalize)
+      }
```

Comment:
What about all the other types?

Reply:
Fair enough.

I'd import `git` here too as it's the one other (well) supported codec on go-ipfs side (and would be really handy to have for https://github.com/ipfs-shipyard/IGiS which can't use `window.ipfs` from api now)

Reply:
Is it OK if I I open an issue for it and do it in a separate PR? That' also something newcomers can work on.

Reply:
@diasdavid Sounds like a test for that is missing in interface-ipfs-core then.

It will work once js-ipld is used.

Reply:
Summary of this thread: everything mentioned here is solved if js-ipld is used. Hence I propose merging this change as is currently is and open a new issue (I'll do that) to use js-ipld instead.

Reply:
Sounds good my IPLD captain! 🚢